### PR TITLE
Add "hermit manifest create <url>".

### PR DESCRIPTION
**Pull request by alecthomas**
This will attempt to infer a manifest from the given download artefact.
It is not super comprehensive but should at the very least check that
there are valid URLs. Only works on GitHub.

```console
$ hermit manifest create https://github.com/mattn/goreman/releases/download/v0.3.7/goreman_v0.3.7_darwin_amd64.zip
// Relative glob from $root to individual terminal binaries.
binaries = []

// Platform-specific configuration. <attr> is a set regexes that must all match against one of CPU, OS, etc..
platform "darwin" "amd64" {
  // URL for source package.
  source = "https://github.com/mattn/goreman/releases/download/v${version}/goreman_v${version}_${os}_${arch}.zip"
}

// Platform-specific configuration. <attr> is a set regexes that must all match against one of CPU, OS, etc..
platform "darwin" "arm64" {
  // URL for source package.
  source = "https://github.com/mattn/goreman/releases/download/v${version}/goreman_v${version}_${os}_amd64.zip"
}

// Platform-specific configuration. <attr> is a set regexes that must all match against one of CPU, OS, etc..
platform "linux" "amd64" {
  // URL for source package.
  source = "https://github.com/mattn/goreman/releases/download/v${version}/goreman_v${version}_${os}_${arch}.tar.gz"
}

// Human readable description of the package.
description = "foreman clone written in go language"

// Definition of and configuration for a specific version.
version "0.3.7" {
  // Automatically update versions.
  auto-version {
  }
}
```

### CODE REVIEW

File: manifest/infer.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+)
+
+// InferFromArtefact attempts to infer a Manifest from a package artefact.
+//
+// "url" should be the full URL to a package artefact.
+//
+//     https://github.com/protocolbuffers/protobuf-go/releases/download/v1.27.1/protoc-gen-go.v1.27.1.darwin.amd64.tar.gz
+//
+// "version" may be specified if it cannot be inferred from the URL.
+func InferFromArtefact(p *ui.UI, httpClient *http.Client, ghClient *github.Client, url, version string) (*Manifest, error) {
```
Comment (jvmakine) [inline manifest/infer.go:32]: This function feels a bit long, maybe it could be split up?

**Review by jvmakine — APPROVED** [review]
Nice! Just a few comments / suggestions. Up to you if you think they are useful.

File: app/main.go [hunk 2]
```diff
@@ -166,7 +170,12 @@ func Main(config Config) {
 		log.Printf("%s: %s", userConfigPath, err)
 	}
 
-	hermitHelp := help + "\n\nConfiguration format for ~/.hermit.hcl:\n    " + strings.Join(strings.Split(userConfigSchema, "\n"), "\n    ")
+	githubToken := os.Getenv("HERMIT_GITHUB_TOKEN")
+
+	hermitHelp := help
+	hermitHelp += "\n\nConfiguration format for ~/.hermit.hcl:\n"
+	hermitHelp += "    " + strings.Join(strings.Split(userConfigSchema, "\n"), "\n    ")
+	hermitHelp += "\nHERMIT_GITHUB_TOKEN can be set to retrieve private GitHub release assets."
```
[the remark this answers is not in the document]
Comment (alecthomas) [reply app/main.go:178]: Rule of 3!

File: github/api.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
+// Package github implements a client for GitHub that includes the minimum set
+// of functions required by Hermit.
+package github
```
[the remark this answers is not in the document]
Comment (alecthomas) [reply github/api.go:3]: I have, definitely, but it's actually quite massive. Perhaps if we end up needing some more functionality, but at this stage I'm fine with ~100 lines of code.

File: manifest/infer.go (hunk 1 above)
Comment (alecthomas) [reply manifest/infer.go:32]: Yes good idea, I had intended to but forgot!

**Review by jvmakine — APPROVED** [review]

**Pull request merged**